File: remarkable/src/s57.cpp
// SPDX-FileCopyrightText: 2024-2026 Jeremy Collins <jeremy.collins@beetlebug.org>
// SPDX-License-Identifier: MIT

#include "s57.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonValue>
#include <QStringList>
#include <QVariantMap>
#include <cmath>

namespace S57 {
namespace {

// The attribute names that carry something to read.
const QStringList kInformational = {
    QStringLiteral("INFORM"),  QStringLiteral("NINFOM"), QStringLiteral("TXTDSC"),
    QStringLiteral("NTXTDS"),  QStringLiteral("PICREP"), QStringLiteral("fileReference"),
    QStringLiteral("text"),
};

const QStringList kFileRefNames = {
    QStringLiteral("TXTDSC"), QStringLiteral("NTXTDS"),
    QStringLiteral("PICREP"), QStringLiteral("fileReference"),
};

QString textOf(const QJsonValue& v) {
    switch (v.type()) {
    case QJsonValue::String:
        return v.toString();
    case QJsonValue::Double: {
        const double d = v.toDouble();
        // Whole numbers print without a trailing .0, as NSNumber.stringValue does.
        if (d == std::floor(d) && std::fabs(d) < 1e15)
            return QString::number(qlonglong(d));
        return QString::number(d);
    }
    case QJsonValue::Bool:
        return v.toBool() ? QStringLiteral("true") : QStringLiteral("false");
    case QJsonValue::Null:
        return QStringLiteral("null");
    default:
        return QString();
    }
}

// Walks the payload exactly as the Swift does: an object or an array with a
// name becomes a heading, and its parts indent under it.
void append(const QJsonValue& node, const QString& name, bool hasName, int depth,
            QVector<Row>& rows) {
    if (node.isObject()) {
        if (hasName)
            rows.append({name, QString(), depth});
        const QJsonObject obj = node.toObject();
        // QJsonObject keys are already sorted, matching `object.keys.sorted()`.
        for (const QString& key : obj.keys())
            append(obj.value(key), key, true, hasName ? depth + 1 : depth, rows);
        return;
    }
    if (node.isArray()) {
        if (hasName)
            rows.append({name, QString(), depth});
        const QJsonArray list = node.toArray();
        for (const QJsonValue& item : list)
            append(item, QString(), false, depth + 1, rows);
        return;
    }
    rows.append({hasName ? name : QString(), textOf(node), depth});
}

} // namespace

bool Row::fileReference() const {
    return kFileRefNames.contains(name) && !value.isEmpty();
}

bool Row::isPicture() const {
    const QString lower = value.toLower();
    return lower.endsWith(QStringLiteral(".tif")) || lower.endsWith(QStringLiteral(".tiff")) ||
           lower.endsWith(QStringLiteral(".jpg")) || lower.endsWith(QStringLiteral(".jpeg")) ||
           lower.endsWith(QStringLiteral(".png"));
}

QVector<Row> attributes(const QString& json) {
    QVector<Row> rows;
    if (json.isEmpty())
        return rows;
    const QJsonDocument doc = QJsonDocument::fromJson(json.toUtf8());
    if (doc.isNull())
        return rows;
    append(doc.isArray() ? QJsonValue(doc.array()) : QJsonValue(doc.object()), QString(), false, 0,
           rows);
    return rows;
}

QVariantList attributeRows(const QString& json) {
    QVariantList out;
    const QVector<Row> rows = attributes(json);
    out.reserve(rows.size());
    for (const Row& r : rows) {
        QVariantMap m;
        m[QStringLiteral("name")] = r.name;
        m[QStringLiteral("value")] = r.value;
        m[QStringLiteral("depth")] = r.depth;
        m[QStringLiteral("fileReference")] = r.fileReference();
        m[QStringLiteral("isPicture")] = r.isPicture();
        out.append(m);
    }
    return out;
}

QString plainText(const QString& cls, const QString& chart, const QString& json) {
    QString text = cls + QStringLiteral("  ") + chart + QStringLiteral("\n");
    for (const Row& r : attributes(json)) {
        const QString indent(r.depth * 2, QLatin1Char(' '));
        text += r.value.isEmpty() ? indent + r.name + QStringLiteral(":\n")
                                  : indent + r.name + QStringLiteral(": ") + r.value +
                                        QStringLiteral("\n");
    }
    return text;
}

bool carriesInformation(const QString& json) {
    for (const Row& r : attributes(json)) {
        if (kInformational.contains(r.name) && !r.value.isEmpty())
            return true;
    }
    return false;
}

} // namespace S57

File: apple/LookoutMarine/Pick/S57.swift
//  S57.swift — the attribute payload, as the cell states it.
//
//  An S-57 cell gives a flat object of acronym and value. S-101 does not: a
//  complex attribute has sub-attributes, so a value can be an object or an
//  array. The rows therefore have a depth, and a complex attribute becomes a
//  heading with its parts indented under it.
//
//  This is the fold under the decoded report, and the clipboard copy, which is
//  how a chart problem gets reported.

import Foundation


/// Attribute payload helpers.
///
/// An S-57 cell gives a flat object of acronym and value. S-101 does not: a
/// complex attribute carries sub-attributes, so a value can be an object or an
/// array. The rows therefore carry a depth, and a complex attribute becomes a
/// heading with its parts indented under it.
enum S57 {
    struct Row: Identifiable {
        let name: String
        let value: String
        let depth: Int
        var id: String { "\(depth)/\(name)/\(value)" }

        /// A cell can point at a text file or a picture beside it, such as
        /// US348MDE.TXT. S-57 names it in TXTDSC, NTXTDS or PICREP; S-101 puts
        /// it in a fileReference. The bake does not carry those files, so the
        /// report states the reference and marks it as a file.
        var fileReference: Bool {
            ["TXTDSC", "NTXTDS", "PICREP", "fileReference"].contains(name)
                && !value.isEmpty
        }

        var isPicture: Bool {
            let lower = value.lowercased()
            return [".tif", ".tiff", ".jpg", ".jpeg", ".png"].contains { lower.hasSuffix($0) }
        }
    }

    static func attributes(of json: String) -> [Row] {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data)
        else { return [] }
        var rows: [Row] = []
        append(root, name: nil, depth: 0, into: &rows)
        return rows
    }

    /// Rows from an already-parsed payload — the envelope's raw half.
    static func rows(of any: Any?) -> [Row] {
        guard let any else { return [] }
        var rows: [Row] = []
        append(any, name: nil, depth: 0, into: &rows)
        return rows
    }

    private static func append(_ node: Any, name: String?, depth: Int, into rows: inout [Row]) {
        switch node {
        case let object as [String: Any]:
            if let name { rows.append(Row(name: name, value: "", depth: depth)) }
            for key in object.keys.sorted() {
                append(object[key] ?? "", name: key, depth: name == nil ? depth : depth + 1, into: &rows)
            }
        case let list as [Any]:
            if let name { rows.append(Row(name: name, value: "", depth: depth)) }
            for item in list {
                append(item, name: nil, depth: depth + 1, into: &rows)
            }
        default:
            rows.append(Row(name: name ?? "", value: text(of: node), depth: depth))
        }
    }

    private static func text(of node: Any) -> String {
        if let s = node as? String { return s }
        if let n = node as? NSNumber { return n.stringValue }
        return String(describing: node)
    }

    /// The attribute names that carry something to read.
    static let informational: Set<String> = [
        "INFORM", "NINFOM", "TXTDSC", "NTXTDS", "PICREP", "fileReference", "text",
    ]

    /// True when the payload holds a note or a reference. It is what keeps a
    /// meta object in the report: M_NPUB carries the chart's cautions, M_QUAL
    /// carries nothing a mariner reads.
    static func carriesInformation(_ json: String) -> Bool {
        attributes(of: json).contains { informational.contains($0.name) && !$0.value.isEmpty }
    }

    /// The report, as plain text for the clipboard: the raw payload, out of
    /// the envelope when there is one — a chart problem is reported with the
    /// cell's own words.
    static func plainText(_ feature: PickFeature) -> String {
        let root = (try? JSONSerialization.jsonObject(with: Data(feature.s57.utf8)))
            as? [String: Any]
        let raw = (root?["report"] != nil ? root?["s57"] : root) as Any?
        var text = "\(feature.cls)  \(feature.chart)\n"
        for row in rows(of: raw) {
            let indent = String(repeating: "  ", count: row.depth)
            text += row.value.isEmpty ? "\(indent)\(row.name):\n"
                                      : "\(indent)\(row.name): \(row.value)\n"
        }
        return text
    }
}
